quixstreams/sources/core/csv.py

```diff
@@ -1,74 +1,81 @@
+            It must return time in milliseconds as `int`.
+            If empty, the current epoch will be used.
+            Default - `None`
+        :param shutdown_timeout: Time in second the application waits for the source to gracefully shut down.
         :param dialect: a CSV dialect to use. It affects quoting and delimiters.
             See the ["csv" module docs](https://docs.python.org/3/library/csv.html#csv-fmt-params) for more info.
             Default - `"excel"`.
-        :param key_deseralizer: a callable to convert strings to key.
-            Default - `str`
-        :param value_deserializer: a callable to convert strings to value.
-            Default - `json.loads`
         """
-        super().__init__(name or path, shutdown_timeout)
         self.path = path
+        self.delay = delay
         self.dialect = dialect
 
-        self._key_deserializer = key_deserializer
-        self._value_deserializer = value_deserializer
+        self.key_extractor = key_extractor
+        self.timestamp_extractor = timestamp_extractor
 
-    def run(self):
-        key_deserializer = self._key_deserializer
-        value_deserializer = self._value_deserializer
+        super().__init__(name=name, shutdown_timeout=shutdown_timeout)
 
+    def run(self):
+        # Start reading the file
         with open(self.path, "r") as f:
+            logger.info(f'Producing data from the file "{self.path}"')
             reader = csv.DictReader(f, dialect=self.dialect)
 
             while self.running:
                 try:
-                    item = next(reader)
+                    row = next(reader)
                 except StopIteration:
                     return
 
-                # if a timestamp column exist with no value timestamp is ""
-                timestamp = item.get("timestamp") or None
-                if timestamp is not None:
-                    timestamp = int(timestamp)
-
-                msg = self.serialize(
-                    key=key_deserializer(item["key"]),
-                    value=value_deserializer(item["value"]),
-                    timestamp_ms=timestamp,
+                # Extract message key from the row
+                message_key = self.key_extractor(row) if self.key_extractor else None
+                # Extract timestamp from the row
+                timestamp = (
+                    self.timestamp_extractor(row) if self.timestamp_extractor else None
                 )
+                # Serialize data before sending to Kafka
+                msg = self.serialize(key=message_key, value=row, timestamp_ms=timestamp)
 
-                self.produce(
-                    key=msg.key,
-                    value=msg.value,
-                    timestamp=msg.timestamp,
-                    headers=msg.headers,
-                )
+                # Publish the data to the topic
+                self.produce(timestamp=msg.timestamp, key=msg.key, value=msg.value)
```

Comment:
Not related to this PR, but would be good if `produce` could just accept `KafkaMessage`...

Reply:
That would reduce the boilerplate, yes.
Maybe we should simply do the serialization inside `produce()`